### ISSUE TITLE
Chaos 0, 2, and 4 Water Bubble Effects (Plus others...)

### ISSUE THREAD

**Issue by supercoolsonic** [PiKeyAr/sadx_dreamcast#202]
I have recently noticed that the water effect inside of Chaos in 3 of the 5 fights against him is different between Dreamcast and DX. Very different, in fact. For Chaos 0, 2, and 4, in the DC version, the water droplets/bubbles (whatever they are) look like they bubble up inside him, from the bottom to the top. This is most easily seen in Chaos 2. Meanwhile, in DX, all of the droplets don't do anything like this at all. They just form on Chaos for a brief time, like little pimples. I was wondering if it's possible to restore these effects, or if they even have any coding left in DX. I should note that this effect is still intact in DX when Chaos 2 is bouncing around the room in his ball form.

Two slightly related other things I noticed, particularly about Chaos 4:
Chaos 4 is meant to have water ripples on the surface of the water as he moves around:
![waterripple](https://user-images.githubusercontent.com/26051785/51367850-217f4300-1ab2-11e9-8e42-fe243ad1c057.PNG)
But they don't show up in the DX version. I'm not sure if they're still actually there but aren't appearing due to transparency stuff. Edit: I've discovered that it still exists and spawns correctly, but can't be seen due to transparency issues. Edit2: They don't show up even while using the dx11 mod, which is confusing to me. They only way I can get them to show is by disabling their alpha flag, but that of course makes them look ugly.
Edit3: Also also, it seems that the model has glitched graphics in all versions of the game, including Dreamcast, because the model is trying to use a texture outside the range of the texlist. There's only 1 texture in the texture file it uses, and it wants to use the 2nd slot. Maybe it would be worth it, if you can get this appearing correctly, to fix its material ID?

And:
![effect](https://user-images.githubusercontent.com/26051785/51367882-3b208a80-1ab2-11e9-8cc1-5b7012cc4efc.PNG)
The color of the effect that plays before Chaos 4 does his shockwave attack is blue-purple on Dreamcast, but red-yellow for DX. Edit: I've discovered this is caused by material colors. The other part of this attack, the wave that can hurt you, is also affected by material colors.

Edit2: I have done further research on this, and it turns out this cannot be solved through material color fixing alone. One of the NJS_ACTIONS for the shockwave attack seems to be using the wrong model. This action is located at 11C117C in the PC version. It is set to use NJS_OBJECT 0x011BFFF4, but it seems it should use 11C0F04. I could be wrong about this, but removing the material colors and swapping the models around like I said seems to fix all consistency issues with the attack.

**Comment by PiKeyAr** [comment]
I've been investigating the bubble effect, and here's what I found out:

The code for the effect (found at 0x7ADD30) is shared between Chaos 0, 2 and 4. The bubbles have multiple "states" (as seen in the switch statement) they go through before they disappear. The final position of the bubble is set using an njTranslateV call at the end of the sub right before drawing the sprite, and right before that it calculates the vector used for that call. Messing around the calculations made the bubbles move erratically, so it's not a queue or particle problem. It appears that the code doing all calculations is working (I changed random things within each switch case and it cause different behavior), but somehow the end result for that vector is always 0, 0, 0 so the bubbles don't move. There is some stuff looking "unnatural", for example there's a subtraction using a negative value (fsub -0.5), which isn't "incorrect" per se but makes me wonder if there's some oversight related to signs or Endianness conversion from the Gamecube or something. The effect works in the Gamecube version. Sadly the version found at 0x550240 (Chaos 2 bouncing as a bubble) is a much simpler effect so it can't be used as a reference even though it's using the same sprite.

**Comment by PiKeyAr** [comment]
Rewrote the original function to replicate vanilla behavior, trying to patch it now
```
struct BUBBLE_LIST
{
	__int16 id;
	Sint8 num;
	Sint8 padding;
	NJS_POINT3* pos;
};


struct BUBBLE_TABLE
{
	unsigned __int16 objnum;
	NJS_POINT3 pos;
	__int16 padding;
	int ang[3];
};


struct BUBBLE_DATA
{
	char object_num;
	char spline_num;
	__int16 padding;
	BUBBLE_LIST* list;
	BUBBLE_TABLE* tbl;
	float scl;
};

DataPointer(BUBBLE_LIST*, ChaosBubbleEffect_BubbleList, 0x3D0DC00);
DataArray(float, SomeScaleOffsetArray, 0x38E2020, 16);
DataArray(char, byte_3D0DBC0, 0x3D0DBC0, 64);
DataPointer(float, min_fluctuation, 0x3D0DB90);
DataPointer(NJS_SPRITE, ChaosBubbleEffect_Sprite, 0x3D0DB9C);
DataPointer(NJS_ARGB, ChaosBubbleEffect_Color, 0x38E200C);
DataPointer(NJS_MATRIX_PTR, bubble_matrix, 0x3D0DB88);
DataPointer(ObjectMaster, ChaosBubbleEffect_ObjectMaster, 0x3D0DB94);

void __cdecl actBubble(ObjectMaster* a1)
{
	BUBBLE* v1; // esi
	int v2; // edi
	NJS_VECTOR* somechange; // ecx
	float finchange; // ST1C_4
	char spline_id;
	NJS_VECTOR* spline1; // rax
	NJS_VECTOR* spline2; // eax
	double t_subtracted; // st7
	double v18; // st7
	Float spiral_sine; // ST1C_4
	double spiral_cosine; // st7
	double scale_change; // st6
	NJS_VECTOR v; // [esp+Ch] [ebp-Ch]
	int tmax; // [esp+1Ch] [ebp+4h]
	float v28; // [esp+1Ch] [ebp+4h]
	float scale_swing; // [esp+1Ch] [ebp+4h]

	v1 = (BUBBLE*)a1->UnknownB_ptr;
	v2 = ChaosBubbleEffect_BubbleList[v1->id].id;
	switch (v1->mode)
	{
	case 0:
		v1->t = (double)rand() * 0.000030517578f * 0.15f - 0.5f;
		somechange = &v1->spline[ChaosBubbleEffect_BubbleList[v1->id].num];
		finchange = fabs(somechange[-1].y - v1->spline->y + somechange[-1].x - v1->spline->x);
		v1->tadd = (rand() * 0.000030517578 * 0.18f + 0.08f) / finchange * min_fluctuation;
		tmax = ChaosBubbleEffect_BubbleList[v1->id].num - 3;
		v1->timer = 15;
		v1->tmax = tmax + 1.25f;
		v1->spiral_add = (Uint32)(((double)rand() * 0.000030517578f * 0.8f + 0.2f) * 30.0f * 65536.0f * 0.002777777777777778f);
		v1->swing = (double)rand() * 0.000030517578f * 0.1f + 0.05f;
		v1->size = SomeScaleOffsetArray[int(rand() * 0.000030517578f * 16.0f)];
		v1->scale_x = 0.1f;
		v1->mode = 1;
		goto swing;
	case 1:
	swing:
		v1->timer -= 1;
		v1->t += v1->tadd;
		v1->scale_x = v1->scale_x + 0.06f;
		if (!v1->timer)
		{
			v1->mode = 2;
		}
		goto render;
	case 2:
		v1->t += v1->tadd;
		if (v1->t >= v1->tmax)
		{
			v1->timer = 8;
			v1->mode = 3;
		}
		v1->spiral += v1->spiral_add;
		v1->scale_x = 0.85f;
		goto render;
	case 3:
		v1->timer -= 1;
		v1->scale_x -= 0.1125f;
		if (v1->timer)
		{
			goto render;
		}
		CheckThingButThenDeleteObject(a1);
		break;
	default:
	render:
		if (!MissedFrames && !((_DWORD)ChaosBubbleEffect_ObjectMaster.Next[1].MainSub & 4) && !byte_3D0DBC0[v2])
		{
			spline_id = (unsigned __int64)(v1->t - -0.5);
			double timer_decrease = (double)(signed int)spline_id;
			spline1 = &v1->spline[spline_id];
			t_subtracted = v1->t - timer_decrease;
			v28 = (1.0f - t_subtracted) * (1.0 - t_subtracted) * 0.5f;
			v18 = t_subtracted * t_subtracted * 0.5f + (1.0f - t_subtracted) * t_subtracted + 0.5f + v28;
			v.x = v18 * spline1->x;
			v.y = v18 * spline1->y;
			v.z = v18 * spline1->z;
			scale_swing = v1->swing * v1->scale_x;
			spiral_sine = njSin(v1->spiral);
			spiral_cosine = njCos(v1->spiral);
			v.x = spiral_sine * scale_swing + v.x;
			v.z = scale_swing * spiral_cosine + v.z;
			scale_change = min_fluctuation * v1->size;
			ChaosBubbleEffect_Sprite.sx = (spiral_sine * 0.3f + v1->scale_x)* scale_change;
			ChaosBubbleEffect_Sprite.sy = (spiral_cosine * 0.3f + v1->scale_x) * scale_change;
			njColorBlendingMode(0, NJD_COLOR_BLENDING_SRCALPHA);
			njColorBlendingMode(NJD_DESTINATION_COLOR, NJD_COLOR_BLENDING_INVSRCALPHA);
			SetMaterialAndSpriteColor(&ChaosBubbleEffect_Color);
			njPushMatrix((NJS_MATRIX_PTR)bubble_matrix + 16 * ChaosBubbleEffect_BubbleList[v1->id].id);
			njTranslateV(0, &v);
			njDrawSprite3D_Queue(&ChaosBubbleEffect_Sprite,	0, NJD_SPRITE_ALPHA | NJD_SPRITE_SCALE | NJD_SPRITE_COLOR, QueuedModelFlagsB_SomeTextureThing);
			njPopMatrix(1u);
			ClampGlobalColorThing_Thing();
			njColorBlendingMode(0, NJD_COLOR_BLENDING_SRCALPHA);
			njColorBlendingMode(NJD_DESTINATION_COLOR, NJD_COLOR_BLENDING_INVSRCALPHA);
		}
		break;
	}
}
```

**Comment by PiKeyAr** [comment]
Cleaned up version of the function with additions from a Gamecube decompilation, almost working:
```void __cdecl actBubble(ObjectMaster* a1)
{
	BUBBLE* v1; // esi
	int v2; // edi
	int spline_id;
	float ampl; // ST1C_4
	float spiral_sine; // ST1C_4
	float spiral_cosine; // st7
	float scale_change; // st6
	float scale_swing; // [esp+1Ch] [ebp+4h]
	float f3;
	float f8;
	float f7;
	NJS_VECTOR v; // [esp+Ch] [ebp-Ch]
	int tmax; // [esp+1Ch] [ebp+4h]

	v1 = (BUBBLE*)a1->UnknownB_ptr;
	v2 = ChaosBubbleEffect_BubbleList[v1->id].id;
	switch (v1->mode)
	{
	case 0:
		v1->t = (double)rand() * 0.000030517578f * 0.15f - 0.5f;
		ampl = fabs(v1->spline[ChaosBubbleEffect_BubbleList[v1->id].num - 1].y - v1->spline[0].y + v1->spline[ChaosBubbleEffect_BubbleList[v1->id].num - 1].x - v1->spline[0].x);
		v1->tadd = (rand() * 0.000030517578 * 0.18f + 0.08f) / ampl * min_fluctuation;
		tmax = ChaosBubbleEffect_BubbleList[v1->id].num - 3;
		v1->timer = 15;
		v1->tmax = tmax + 1.25f;
		v1->spiral_add = (Uint32)(((double)rand() * 0.000030517578f * 0.8f + 0.2f) * 30.0f * 65536.0f * 0.002777777777777778f);
		v1->swing = (double)rand() * 0.000030517578f * 0.1f + 0.05f;
		v1->size = SomeScaleOffsetArray[int(rand() * 0.000030517578f * 16.0f)];
		v1->scale_x = 0.1f;
		v1->mode = 1;
		goto swing;
	case 1:
	swing:
		v1->timer -= 1;
		v1->t += v1->tadd;
		v1->scale_x = v1->scale_x + 0.06f;
		if (!v1->timer)
		{
			v1->mode = 2;
		}
		goto render;
	case 2:
		v1->t += v1->tadd;
		if (v1->t >= v1->tmax)
		{
			v1->timer = 8;
			v1->mode = 3;
		}
		v1->spiral += v1->spiral_add;
		v1->scale_x = 0.85f;
		goto render;
	case 3:
		v1->timer -= 1;
		v1->scale_x -= 0.1125f;
		if (v1->timer)
		{
			goto render;
		}
		CheckThingButThenDeleteObject(a1);
		break;
	default:
	render:
		if (!MissedFrames && !((_DWORD)ChaosBubbleEffect_ObjectMaster.Next[1].MainSub & 4) && !byte_3D0DBC0[v2])
		{
			spline_id = (int)(v1->t + 0.5f);
			f8 = v1->t - (float)spline_id;
			f7 = 1.0f - f8;
			f3 = f8 * f8;
			f7 = f7 * f8 + 0.5f;
			f8 = 0.5f * f3;
			f3 = 0.5f * f7 * f7;
			v.x = v1->spline[spline_id].x * f8 + v1->spline[spline_id].x * f3 + v1->spline[spline_id + 1].x * f7;
			v.y = v1->spline[spline_id].y * f8 + v1->spline[spline_id].y * f3 + v1->spline[spline_id + 1].y * f7;
			v.z = v1->spline[spline_id].z * f8 + v1->spline[spline_id].z * f3 + v1->spline[spline_id + 1].z * f7;
			scale_swing = v1->swing * v1->scale_x;
			spiral_sine = njSin(v1->spiral);
			spiral_cosine = njCos(v1->spiral);
			v.x = spiral_sine * scale_swing + v.x;
			v.z = scale_swing * spiral_cosine + v.z;
			scale_change = min_fluctuation * v1->size;
			ChaosBubbleEffect_Sprite.sx = (spiral_sine * 0.3f + v1->scale_x)* scale_change;
			ChaosBubbleEffect_Sprite.sy = (spiral_cosine * 0.3f + v1->scale_x) * scale_change;
			njColorBlendingMode(0, NJD_COLOR_BLENDING_SRCALPHA);
			njColorBlendingMode(NJD_DESTINATION_COLOR, NJD_COLOR_BLENDING_INVSRCALPHA);
			SetMaterialAndSpriteColor(&ChaosBubbleEffect_Color);
			njPushMatrix((NJS_MATRIX_PTR)bubble_matrix + 16 * ChaosBubbleEffect_BubbleList[v1->id].id);
			njTranslateV(0, &v);
			njDrawSprite3D_Queue(&ChaosBubbleEffect_Sprite,	0, NJD_SPRITE_ALPHA | NJD_SPRITE_SCALE | NJD_SPRITE_COLOR, QueuedModelFlagsB_SomeTextureThing);
			njPopMatrix(1u);
			ClampGlobalColorThing_Thing();
			njColorBlendingMode(0, NJD_COLOR_BLENDING_SRCALPHA);
			njColorBlendingMode(NJD_DESTINATION_COLOR, NJD_COLOR_BLENDING_INVSRCALPHA);
		}
		break;
	}
}```

**Comment by PiKeyAr** [comment]
Fixed in [d2225b6](https://github.com/PiKeyAr/sadx_dreamcast/commit/d2225b687a8f9560c050cc61e68a1aa8ed10fb3b)
